Feat/federatedquery

SteinerPascal opened the pull request:
@tfrancart 
1. This branch contains a refactor of the query generation. It has grown messy and i refactored it now to a more OOP approach. It should be more readable and clean now. It follows roughly the builder pattern (https://refactoring.guru/design-patterns/builder)

2. I have included the SERVICE keyword for datasources which have the sparqlEndpointUrl defined in their datasources (https://docs.sparnatural.eu/JSON-based-configuration#specifying-the-sparql-service-of-a-datasource). If their is a sparqlEndpointUrl defined it uses this as the endpoint in the service keyword and inserts the SERVICE. It works also with optional& notexists OR you can define optional&notexists *within* the service keyword. 

3. It's now going to be simpler to add another pattern schema to the standard query generation pattern (?classtypeVar rdf:type owl:Class). Preparation for more advanced usages of the SERVICE keyword. 

@tfrancart 
What do you think about it? Especially Nr2. is this okay this way?

SteinerPascal commented:
some visual example where i changed the datasource of countries to cantons in switzerland

![image](https://user-images.githubusercontent.com/49486647/196212672-2864bd8f-d2fd-4f80-b1b8-6702ef182cf2.png)

![image](https://user-images.githubusercontent.com/49486647/196214189-61fb1178-c5ad-4c4d-87e7-352c939b82fc.png)
In this example I'm not sure if that's the behavior we actually want. It now behaves the same way as  if an "optional | noteexists" arrow got clicked. Propbably it would make more sense to only activate the SERVICE when there are values selected from a widget?
Or only top two lines within the SERVICE keyword?

tfrancart commented:
Refactoring of query generation looks great.

But the sparqlEndpointUrl on the datasource had a slightly different meaning, I don't think we can rely on it to generate SERVICE keyword.
The usage of sparqlEndpointUrl is the following :
- my Knowledge Graph relies on values of an external data for a certain property; the actual descriptions and labels of those values are not in my KG and they need to be fetched from another endpoint URL
- but the actual property (what you call the "intersection triple" in the code) lives inside my KG - no need for a SERVICE keyword

This is different from stating that the **property** does not live in my KG, but lives elsewhere.

But this raises the question of the population of the list or autocomplete combined with the SERVICE keyword. I have added a comment here on this : https://github.com/sparna-git/Sparnatural/issues/352#issuecomment-1284222216

In the tests you did above, the queries are correct if we consider they are a simulation of what would actually be a real Service configuration on the property, and not on the datasource; this can't stay like this.

SteinerPascal commented on src/sparnatural/generators/SelectBuilder.ts:
```diff
@@ -0,0 +1,147 @@
+import IntersectionBuilder from "./rdf/IntersectionBuilder";
+import SparqlFactory from "./SparqlFactory";
+
+export default class SelectBuilder{
```
Yes, that actually makes a lot of sense. I'll open a new issue on this.

SteinerPascal commented:
I'm not sure if I can agree with the usage of the current sparqlEndpointURL. Let me explain my point of view to this matter:

In your definition you say that the KG relies on external values for a certain property. But once the external value is chosen we will appearantly find it in our KG since it will be inserted into the query. 

Scenario: I would like to know all museums in certain countries. The countries are populated from the ListWidget. The datasource for the ListWidget is different from the default endpoint. Meaning the labels are located on a different KG than "my" KG.
![image](https://user-images.githubusercontent.com/49486647/196768953-34c44e4f-5a86-402e-9ad4-5a8f11ce024c.png)

Now here is what i don't understand: Why would i choose a different sparqEndpointURL if the value in the VALUES clause is within my KG. I would rather have the list widget populated from my KG (since the value ` <http://fr.dbpedia.org/resource/Afrique_du_Sud>` appearently is in my KG. otherwise the query would not return any results. Using a different sparqlEndpoint only to generate the labels for the listwidget seems a bit weird? Can you provide me with a usecase of yours?

But I agree that we don't necessarly need to map the SERVICE keyword to the datasource property.

tfrancart commented:
> Using a different sparqlEndpoint only to generate the labels for the listwidget seems a bit weird? Can you provide me with a usecase of yours?

Yes, the sparqlEndpointUrl is indeed only to fetch the labels for the list widgets. It is attached to the datasource, so it really means : "execute the SPARQL query of the datasource on this SPARQL endpoint". I had 2 use-cases:
  1. one was a system allowing to describe persons with their skills, and the skill was searched and fecth from DBPedia, referring to a DBPedia URI
  2. one is the BNF (French National Library) referring to the URI of languages in LCSH (library of Congress Subject Headings)

When you think about it, this is a common pattern when doing Linked Data : point to URIs of entities elsewhere, and leave it to the client (in this case Sparnatural) to aggregate my KG with the other datasources.

So:
1. If I have a sparqlEndpointUrl on the datasource, this does not imply the use of a SERVICE keyword to query the property(-ies) using this datasource
2. But if I have a service configured on a property, then the datasource of this property necessarily need to be executed on this service

SteinerPascal commented:
Okay i can see your reasoning. I also agree to the two rules you mentioned.  I will change the usage of the SERVICE keyword to the usage of the property.

SteinerPascal commented:
changes:
- added a lot of cleanup/renaming
- cleaned package.json of unused depenedencies
- upgraded old dependencies causing npm vulnerabilities
- prepared package.json for npm publishing
- added code for service config

TODO:
-getting OK for issue #137 from @tfrancart , then ready to merge
